tokio/src/io/async_fd.rs

```diff
@@ -641,22 +753,47 @@ impl<T: AsRawFd> Drop for AsyncFd<T> {
 }
 
 impl<'a, Inner: AsRawFd> AsyncFdReadyGuard<'a, Inner> {
-    /// Indicates to tokio that the file descriptor is no longer ready. The
-    /// internal readiness flag will be cleared, and tokio will wait for the
+    /// Indicates to tokio that the file descriptor is no longer ready. All
+    /// internal readiness flags will be cleared, and tokio will wait for the
     /// next edge-triggered readiness notification from the OS.
     ///
+    /// This function is commonly used with guards returned by [`AsyncFd::readable`] and
+    /// [`AsyncFd::writable`].
+    ///
     /// It is critical that this function not be called unless your code
     /// _actually observes_ that the file descriptor is _not_ ready. Do not call
     /// it simply because, for example, a read succeeded; it should be called
     /// when a read is observed to block.
-    ///
-    /// [`drop`]: method@std::mem::drop
     pub fn clear_ready(&mut self) {
         if let Some(event) = self.event.take() {
             self.async_fd.registration.clear_readiness(event);
         }
     }
 
+    /// Indicates to tokio that the file descriptor no longer has a specific readiness.
+    /// The internal readiness flag will be cleared, and tokio will wait for the
+    /// next edge-triggered readiness notification from the OS.
+    ///
+    /// This function is useful in combination with the [`AsyncFd::ready`] method when a
+    /// combined interest like `Interest::READABLE | Interest::WRITABLE` is used.
+    ///
+    /// It is critical that this function not be called unless your code
+    /// _actually observes_ that the file descriptor is _not_ ready for the provided `Ready`.
+    /// Do not call it simply because, for example, a read succeeded; it should be called
+    /// when a read is observed to block. Only clear the specific readiness that is observed to
+    /// block. For example when a read blocks when using a combined interest,
+    /// only clear `Ready::READABLE`.
+    pub fn clear_ready_exact(&mut self, ready: Ready) {
```

Comment:
It looks like this fn doesn't check if the supplied `ready` is a superset of the event's readiness. If it is, it seems that `self.event` should be cleared (as in `clear_ready()`).

* Could you add a test for this method covering these cases?
* Could you include a doc example as well?

I don't *love* the name `clear_ready_exact`,  but I also don't have a better idea, so I'm OK sticking with it. I do like that it starts with the `clear_ready` prefix, though.

I asked ChatGPT for some ideas, I kind of like `clear_ready_matching` or `clear_ready_with`.

ChatGPT's answer:

> Choosing a method name can often be a matter of personal style and the idiomatic conventions in a codebase, but here are some suggestions for the method name:
> 
> 1. `clear_ready_intersection`: This name clearly communicates that the method clears the intersection of readiness.
> 
> 2. `clear_ready_matching`: This implies that only the readiness bits that match those of the argument will be cleared.
> 
> 3. `clear_ready_by`: This could suggest that the readiness is cleared "by" the criteria of the given `Ready` instance.
> 
> 4. `clear_ready_with`: The "with" here could suggest that the clearing action is done in accordance with the `Ready` instance that's passed.
> 
> 5. `clear_ready_bits`: This is more general, but the "bits" could indicate that you're passing another bitfield to determine what's cleared.
> 
> 6. `clear_ready_shared`: This name would indicate that it's clearing readiness that's shared between the socket and the `Ready` instance.
> 
> 7. `clear_ready_conditional`: This might suggest that there's some condition (the `Ready` argument) that determines which readiness is cleared.
> 
> 8. `clear_ready_overlap`: The word "overlap" denotes the intersection concept.
> 
> Remember, it's also important to document the behavior of this method clearly in comments or other forms of documentation so that future developers understand how it works.

Reply:
> I don't love the name clear_ready_exact, but I also don't have a better idea

fully agree! in context, I think I like `clear_ready_matching` best:

    ///                 Err(ref e) if e.kind() == io::ErrorKind::WouldBlock => {
    ///                     guard.clear_ready_exact(Ready::READABLE);
    ///                     continue;
    ///                 }

    ///                 Err(ref e) if e.kind() == io::ErrorKind::WouldBlock => {
    ///                     guard.clear_ready_matching(Ready::READABLE);
    ///                     continue;
    ///                 }

    ///                 Err(ref e) if e.kind() == io::ErrorKind::WouldBlock => {
    ///                     guard.clear_ready_with(Ready::READABLE);
    ///                     continue;
    ///                 }

For testing, I added some basic tests. What is not (yet) tested is that the event is fully cleared. This is tricky, because clearing the event is not observable from the outside. The same is actually true for the existing `clear_ready` function. I'm also not sure it's that relevant. it's an optimization, but not required for correctness.

It is hard to come up with a succinct example here. This function only makes sense with a combined interest, or when using the priority interest but it is linux-only. So, I can't really come up with anything better than just copy-pasting the example from `ready`. I'm happy to do that, but it would duplicate that example 4 times in total (ready, ready_mut, and clear_ready_matching for the two guard types. There is risk of introducing inconsistencies. Anyway, thoughts?

Reply:
the `clear_ready_matching` functions now have the example from their matching `ready` function as an example.